docs/learn/config.md
---
title: Scully Config
published: true
lang: en
position: 100
---

# Scully Config

<div class="docs-link_table">
  <a class="view-in-repo" href="https://github.com/scullyio/scully/blob/main/libs/scully/src/lib/utils/interfacesandenums.ts"></a>
</div>

## Overview

The central part of a Scully project is the `scully.<projectname>.config.ts` file. This file exports the Scully build configuration for an application.

If you are new to Scully, it is recommended to read the [Getting Started](docs/learn/getting-started/requirements/) documentation.

The `scully.<projectname>.config.ts` file's structure is shown below:

## Interface

```typescript
export interface ScullyConfig {
  /** is this a bare project (without angular.json?) */
  bareProject?: boolean;
  /** the name of the project we are using. Provided by Scully itself */
  projectName?: string;
  /** the folder where project is. Can be any off the projects in a repo, read from angular.json */
  projectRoot?: string;
  /** the folder where the project sources resides, read from angular.json */
  sourceRoot?: string;
  /** Array with string ID's of the content-renderers that will be run on all routes */
  defaultPostRenderers?: string[];
  /** the root of the project (where angular.json lives) */
  homeFolder?: string;
  /** the destination of the Scully generated files */
  outDir?: string;
  /** the place where distribution files of the project are. Should be a subfolder of dist. */
  distFolder?: string;
  /** transferState only inlined into page, and not written into separate data.json */
  inlineStateOnly?: boolean;
  /** Set what is what is written to the logfile, defaults to warnings and errors */
  logFileSeverity?: LogSeverity;
  /** routes that need additional processing have their configuration in here */
  routes: RouteConfig;
  /** routes that are in the application but have no route in the router */
  extraRoutes?: string | string[] | Promise<string[] | string>;
  /** Port-number where the original application is served */
  appPort?: number;
  /** Boolean that determines saving of site-tumbnails files */
  thumbnails?: boolean;
  /** port-number where the Scully generated files are available */
  staticport?: number;
  /** port for the live reload service */
  reloadPort?: number;
  /** optional proxy config file, uses the same config file as the CLI */
  proxyConfig?: string;
  /** optional launch-options for puppeteer */
  puppeteerLaunchOptions?: LaunchOptions;
  /** hostname to use for local server, defaults to `localhost` */
  hostName?: string;
  /** optional hostURL, if this is provided, we are going to use this server instead of the build-in one. */
  hostUrl?: string;
  /** optional guessParserOptions, if this is provided we are going to pass those options to the guess parser. */
  guessParserOptions?: GuessParserOptions;
  /** the maximum of concurrent puppeteer tabs open. defaults to the available amounts of cores */
  maxRenderThreads?: number;
  /** the resource types to ignore when generating pages via Puppeteer */
  ignoreResourceTypes?: ResourceType[];
  /** how to handle 404 in Scully server */
  handle404?: string;
}
```

## Properties

The `ScullyConfig` interface provides parameters for configuring how Scully works in a project.

#### `projectRoot?:` _`string`_

The project's from which Scully generates the static content.

#### `homeFolder?:` _`string`_

A reference to the Angular project's root folder.  
This property is for internal use, and it defaults to the angular.json file's location.

#### `outDir?:` _`string`_

The folder's path where Scully leaves the statics files.  
This should not be the same as the `distFolder`.  
The default path is:

```URL
./dist/static
```

#### `distFolder?:` _`string`_

Path to the Angular application's dist folder.  
Scully takes the `angular.json` file's default path, and will use this folder during rendering.  
This option can be modified according to your needs.

#### `logFileSeverity:` _`LogSeverity`_

Determines what of the Scully output will be written into the `scully.log` file in the root of the project.

| option | result                        |
| ------ | ----------------------------- |
| `0`    | Logs everything               |
| `1`    | Logs warnings and errors only |
| `2`    | Logs errors only              |

#### `routes:` _`RouteConfig`_

Scully has two types of routes, **unhandled routes** and **handled routes**:

##### Unhandled Routes

Routes with dynamic data. This are the routes as you would use them inside your app. Those routes can come from the automated route discovery, or from the extraRoutes property in the `scully.<projectname>.config.ts`. Eg:

```URL
/foo/:id
```

All unhandled routes with dynamic data need to be handled through plugins. When there is a route with dynamic data that has no configuration in the configs routes, it will be logged to screen and skipped during processing.

**This means there will be NO STATIC FILES for ROUTES which HAVE DYNAMIC DATA but NO CONFIG**

For more information about router plugins read the [Plugins](/docs/learn/plugins/overview) documentation.

##### Handled Routes

Routes with static params. Eg:

```URL
/foo/1
```

#### `extraRoutes?:` _`string | string[] | Promise<string[] | string>`_

Allows developers to add an array of unhandled routes. These routes can exist in AngularJS, React, or any other framework.

```typescript
extraRoutes: [
  '/foo/:id',
  new Promise('/bar/:barId'),
  new Promise(['/foo/:fooId', '/bar/:id']),
];
```

#### `appPort?:` _`number`_

Scully provides a server to to render the Angular application.

Configure the port where the Angular application runs.

The default port is: `1864`

#### `staticPort?:` _`number`_

Similar to [_`appPort`_](#appport-number), `staticport` provides a server to to render the static files compiled by Scully.

The default port is: `1668`

#### `proxyConfig?:` _`string`_

Takes a relative filename for a proxy config file.

For more details look at [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware)  
Scully uses the same config format as [webpackDevServer](https://webpack.js.org/configuration/dev-server/#devserverproxy)  
This is an optional property, and it is also used by the [Angular CLI](https://angular.io/guide/build#proxying-to-a-backend-server)

This can also be provided with the `--proxy filname` command line flag

#### `puppeteerLaunchOptions?:` _`LaunchOptions`_

If the application is in a restricted environment, puppeteer's default options may not work. In that case,
this option can be overwrite with settings that match a specific environment.

**A word of warning,** some settings might interfere with the way Scully is working, creating inaccurate results.  
Read about [puppeteerlaunchoptions](https://pptr.dev/#?product=Puppeteer&version=v2.0.0&show=api-puppeteerlaunchoptions) for more information.

#### `hostName?:` _`string`_

Allows to set a different name for the `localhost`.

#### `hostUrl?:` _`string`_

Connects your application to a different host. This is useful when using your own server.

#### `guessParserOptions?:` _`GuessParserOptions`_

The`guessParserOptions` that get passed to the `guess-parser` library.

Currently, the only supported property is `excludedFiles`, and it excludes files from the `guess-parser` route discovery process.

#### `ignoreResourceTypes?:` _`ResourceType[]`_

The `ignoreResourceTypes` array that get passed to the `puppeteerRenderPlugin`.  
Any `ResourceType` that is listed here will be ignored by the Puppeteer instance rendering the requested page.  
For example if you add `image` and `font` all requests to images and fonts loaded on your pages will be ignored.

#### `handle404?:` _`string`_

How routes which are **not**provided in the application, are handled by the Scully server.  
When the server gets a request for a route (file) that does not exist on the file-system, this option amends how that route is handled.

| option         | result                                                       |
| -------------- | ------------------------------------------------------------ |
| `""` (default) | Will render a 404 page, and raise a waring during rendering. |
| `index`        | Will render the `index.html` from the dist root folder.      |
| `baseOnly`     | Will use express route matcher on unhandled routes only.     |
| `404`          | Will render the `404.html` from the dist root folder.        |
| `none`         | Will leave it up to the express software layer.              |


docs/routeParameters.md
---
title: Route Parameters & Scully
order: 190
lang: en
slug: routeParameters
---

# Route Parameters & Scully

If you run Scully and the following warning is displayed, you need to teach Scully how to use the project's route parameters.

```bash
No configuration for route `/user/:userId` found. Skipping
```

The above error is given because Scully does not know all the possible values for `:userId`. Teach Scully how to get the list of `:userId`s from your app. Scully can turn `/user/:userId` into a list of meaningful pre-renderable routes like so:

```bash
/user/1
/user/2
/user/3
...
/user/100
```

Even small Angular projects have routes that contain route parameters. To stop Scully from skipping these routes, configure a route plugin. Route plugins teach Scully how to fetch data and merges it into routes using parameters.

The easiest way to understand route plugin is by understanding the `jsonPlugin`. It simply fetches data from any API that you specify, and it returns a list of properties that can be used to replace the route parameter. Checkout the [jsonPlugin docs](./plugin/jsonPlugin.md) to see an example of how easy this configuration is.


CHANGELOG.md
#### 1.0.0-beta.0 (2020-07-17)

##### Chores

- **deps:**
  - bump lodash from 4.17.15 to 4.17.19 in /libs/scully ([#750](https://github.com/scullyio/scully/pull/750)) ([7d3f7188](https://github.com/scullyio/scully/commit/7d3f718826c101f8e37d4f0d2ab82a6c56be1fa9))
  - bump npm-registry-fetch in /libs/scully-schematics ([#707](https://github.com/scullyio/scully/pull/707)) ([66cf8aeb](https://github.com/scullyio/scully/commit/66cf8aeba41325694dad609b82a3b6e76f8c51a6))
  - bump minimist from 0.0.8 to 1.2.5 in /schematics/scully ([#426](https://github.com/scullyio/scully/pull/426)) ([1af427f7](https://github.com/scullyio/scully/commit/1af427f7d21519e58496057a2d7647c719dcd8ff))
- **monorepo:**
  - update angular version ([#574](https://github.com/scullyio/scully/pull/574)) ([0009e040](https://github.com/scullyio/scully/commit/0009e0401ae3f88efcc6292127d0a5a13dbde2e0))
  - new versions ([#375](https://github.com/scullyio/scully/pull/375)) ([8d8714ba](https://github.com/scullyio/scully/commit/8d8714bae6a6f363c8bb70952ce9a340f432ff5e))
- **scully:**
  - published new version. upackage updates ([#314](https://github.com/scullyio/scully/pull/314)) ([87ffedcd](https://github.com/scullyio/scully/commit/87ffedcd6e85de2b46d88e5d497aa233453967b8))
  - fix ISSUE_TEMPLATEs ([#238](https://github.com/scullyio/scully/pull/238)) ([a7e967d5](https://github.com/scullyio/scully/commit/a7e967d5168d7145d08c653773582b4534b19af9))
  - fix bug ISSUE_TEMPLATE ([#203](https://github.com/scullyio/scully/pull/203)) ([ed778d07](https://github.com/scullyio/scully/commit/ed778d0751896e6e59f6364611f7b236dae51d5b))
- **update:** update package version ([06b35456](https://github.com/scullyio/scully/commit/06b35456396455ae75ad1adca66dc7748b45b250))
- **version update:** version update ([#272](https://github.com/scullyio/scully/pull/272)) ([2ca42bdf](https://github.com/scullyio/scully/commit/2ca42bdf56d7a83c7ec74ff2b8a7fac5a3b09b73))
- **ng-add:** add message for show the angular version support ([#253](https://github.com/scullyio/scully/pull/253)) ([d4c7f0ce](https://github.com/scullyio/scully/commit/d4c7f0ce036a1bc2aa499cc7a771d432969d5c17))
- **version bump and guess test:** bump the version, and add a guess test ([#241](https://github.com/scullyio/scully/pull/241)) ([62cd8014](https://github.com/scullyio/scully/commit/62cd80141d641dfa16b1f55cab4ef844f72bf512))
- **config:** exported scullyConfig ([#222](https://github.com/scullyio/scully/pull/222)) ([6ee2b015](https://github.com/scullyio/scully/commit/6ee2b0159330b234eb0d7ac87b0ad65a176b2e3a))
- **pushed out new packages and updated versions:** pushed out new packages and updated versions ([c7ac1a83](https://github.com/scullyio/scully/commit/c7ac1a83f765a4bbeb9ff56de95842f0c034e034))
- **script:** 🤖 add commit select for commit only the file the pr need ([#158](https://github.com/scullyio/scully/pull/158)) ([2f6f0a12](https://github.com/scullyio/scully/commit/2f6f0a125b6b705c6a4d9bf4a41d4ae61bb6c327))
- **docs:** improve getting started ([497415f2](https://github.com/scullyio/scully/commit/497415f23a5352b2aabfa5b4878ba448268ecc1f))
- **publish:** add np for publish ([398bc85d](https://github.com/scullyio/scully/commit/398bc85d46efd651097894f9f9d52985583a378f))

##### Continuous Integration

- **nodejs.yml:** add pull request to run github actions ([#303](https://github.com/scullyio/scully/pull/303)) ([672ff134](https://github.com/scullyio/scully/commit/672ff134d35b92c4c2cd25ea848671df3001128a))
- **GA:** Add github actions for CI ([#296](https://github.com/scullyio/scully/pull/296)) ([a177a306](https://github.com/scullyio/scully/commit/a177a3066563548eb354f07b546a9cade178864e))

##### Documentation Changes

- landing page feature card on small screens ([#745](https://github.com/scullyio/scully/pull/745)) ([b46282ad](https://github.com/scullyio/scully/commit/b46282adce4a2944339d659348c40e8e45d1faed))
- hide dash on app-code title ([#740](https://github.com/scullyio/scully/pull/740)) ([5fd29e5e](https://github.com/scullyio/scully/commit/5fd29e5e42fb1079c36d60cd9dc84bc84d835d3f))
- replace coming soon note with gitter link ([#626](https://github.com/scullyio/scully/pull/626)) ([#628](https://github.com/scullyio/scully/pull/628)) ([ff453259](https://github.com/scullyio/scully/commit/ff4532597caae129ad9c823aeb2db31dfa278791))
- fix the broken links ([#607](https://github.com/scullyio/scully/pull/607)) ([f309a8fb](https://github.com/scullyio/scully/commit/f309a8fb1c35dd75e2e5be84b9f5b606b9256480))
- update showcase site ([#577](https://github.com/scullyio/scully/pull/577)) ([9a383780](https://github.com/scullyio/scully/commit/9a3837804507dce3338d264af5ebdf2884d7ac0b))
- ✏️ add more descriptive examples to getting-started.md ([#578](https://github.com/scullyio/scully/pull/578)) ([f8ab9842](https://github.com/scullyio/scully/commit/f8ab9842c2afb42dc60ffe885d75b1a7e76d0eec))
- fix href ([#544](https://github.com/scullyio/scully/pull/544)) ([1033e107](https://github.com/scullyio/scully/commit/1033e107f633314fa4eca090657dc9ab6b6ffde7))
- add link to Scully Publish Github action ([#466](https://github.com/scullyio/scully/pull/466)) ([66f0fa8b](https://github.com/scullyio/scully/commit/66f0fa8b12640be5e062b5b92322146d6cf62516))
- add ledge to showcase ([#464](https://github.com/scullyio/scully/pull/464)) ([bb6f5a4a](https://github.com/scullyio/scully/commit/bb6f5a4ad933946f17854e405e8b00563adb16e7))
- remove old `--nw` flag ([#467](https://github.com/scullyio/scully/pull/467)) ([1813af2b](https://github.com/scullyio/scully/commit/1813af2bae4d8bdd248657724992d55d401be945))
- update docs ([#328](https://github.com/scullyio/scully/pull/328)) ([e74e6057](https://github.com/scullyio/scully/commit/e74e6057c86a9c2e60a69bf7e3255836373b8b73))
- fix typos ([#319](https://github.com/scullyio/scully/pull/319)) ([25b33fa4](https://github.com/scullyio/scully/commit/25b33fa4cc0c001487376b99e45fd30eafc4954c))

libs/scully/src/lib/utils/config.ts
/* eslint-disable @typescript-eslint/no-use-before-define */
import { cpus } from 'os';
import { join } from 'path';
import { compileConfig } from './compileConfig';
import { findAngularJsonPath } from './findAngularJsonPath';
import { ScullyConfig } from './interfacesandenums';
import { logError, LogSeverity, logWarn, yellow } from './log';
import { readAngularJson } from './read-anguar-json';
import { validateConfig } from './validateConfig';
export const angularRoot = findAngularJsonPath();
export const scullyConfig: ScullyConfig = {} as ScullyConfig;

export const scullyDefaults: Partial<ScullyConfig> = {
  bareProject: false,
  homeFolder: angularRoot,
  outDir: join(angularRoot, './dist/static/'),
  logFileSeverity: LogSeverity.warning,
  inlineStateOnly: false,
  thumbnails: false,
  maxRenderThreads: cpus().length,
  handle404: '',
  appPort: /** 1864 */ 'herodevs'.split('').reduce((sum, token) => (sum += token.charCodeAt(0)), 1000),
  staticPort: /** 1668 */ 'scully'.split('').reduce((sum, token) => (sum += token.charCodeAt(0)), 1000),
  reloadPort: /** 2667 */ 'scullyLiveReload'.split('').reduce((sum, token) => (sum += token.charCodeAt(0)), 1000),
  hostName: 'localhost',
  defaultPostRenderers: [],
};

const loadIt = async () => {
  const compiledConfig = await compileConfig();
  let angularConfig = {} as any;
  let distFolder = join(angularRoot, './dist');
  let projectConfig: any = {};
  try {
    angularConfig = readAngularJson();
    const defaultProject = compiledConfig.projectName;
    projectConfig = angularConfig.projects[defaultProject];
    distFolder = projectConfig.architect.build.options.outputPath;
    if (distFolder.endsWith('dist') && !distFolder.includes('/')) {
      logError(`Your distribution files are in "${yellow(distFolder)}". Please change that to include a subfolder`);
      process.exit(15);
    }
  } catch (e) {
    // console.log(e);
    logError(`Could not find project "${yellow(compiledConfig.projectName)}" in 'angular.json'.`);
    // process.exit(15);
  }

  if (compiledConfig.hostUrl && compiledConfig.hostUrl.endsWith('/')) {
    compiledConfig.hostUrl = compiledConfig.hostUrl.substr(0, compiledConfig.hostUrl.length - 1);
  }

  // TODO: update types in interfacesandenums to force correct types in here.
  // tslint:disable-next-line: no-unused-expression
  Object.assign(
    scullyConfig,
    /** the default config */
    scullyDefaults,
    /** calculated defaults. */
    {
      sourceRoot: projectConfig.sourceRoot,
      projectRoot: projectConfig.root,
      distFolder,
    }
  ) as ScullyConfig;
  /** activate loaded config */
  await updateScullyConfig(compiledConfig);
  return scullyConfig;
};

/** export the config as a promise, so you can wait for it when you need config during 'boot' */
export const loadConfig = loadIt();

export const updateScullyConfig = async (config: Partial<ScullyConfig>) => {
  /** note, an invalid config will abort the entire program. */
  const newConfig = Object.assign({}, scullyConfig, config);
  if (config.outDir === undefined) {
    logWarn(`The option outDir isn't configured, using default folder "${yellow(scullyConfig.outDir)}".`);
  } else {
    config.outDir = join(angularRoot, config.outDir);
  }
  const validatedConfig = await validateConfig(newConfig as ScullyConfig);
  if (validatedConfig) {
    const mergedRoutes = { ...scullyConfig.routes, ...validatedConfig.routes };
    Object.assign(scullyConfig, config, { routes: mergedRoutes });
  }
};
